web-client/src/models/requests/timeline.ts

```diff
@@ -16,74 +46,161 @@ export interface ITimelineItem extends firebase.firestore.DocumentData {
+
+  get actorSnapshot(): User {
+    return this._actorSnapshot;
+  }
+
+  set actorSnapshot(value: User) {
+    this._actorSnapshot = value;
   }
 
-  get offerSnapshot(): IOffer {
+  @IsNotEmptyObject()
+  @IsOptional()
+  private _offerRef: firestore.DocumentReference<firestore.DocumentData> | null;
+
+  get offerRef(): firestore.DocumentReference<firestore.DocumentData> | null {
+    return this._offerRef;
+  }
+
+  set offerRef(
+    value: firestore.DocumentReference<firestore.DocumentData> | null,
+  ) {
+    this._offerRef = value;
+  }
+
+  @IsObject()
+  @IsOptional()
+  @ValidateNested()
+  private _offerSnapshot: Offer | null;
+
+  get offerSnapshot(): Offer | null {
     return this._offerSnapshot;
   }
 
-  set offerSnapshot(value: IOffer) {
+  set offerSnapshot(value: Offer | null) {
     this._offerSnapshot = value;
   }
 
-  get requestSnapshot(): IRequest {
+  @IsNotEmptyObject()
+  private _requestRef: firestore.DocumentReference<firestore.DocumentData>;
+
+  get requestRef(): firestore.DocumentReference<firestore.DocumentData> {
+    return this._requestRef;
+  }
+
+  set requestRef(value: firestore.DocumentReference<firestore.DocumentData>) {
+    this._requestRef = value;
+  }
+
+  @IsObject()
+  @ValidateNested()
+  private _requestSnapshot: Request;
+
+  get requestSnapshot(): Request {
     return this._requestSnapshot;
   }
 
-  set requestSnapshot(value: IRequest) {
+  set requestSnapshot(value: Request) {
     this._requestSnapshot = value;
   }
 
-  get createdAt(): firebase.firestore.Timestamp {
+  @IsEnum(TimelineItemAction)
+  private _action: TimelineItemAction;
+
+  get action(): TimelineItemAction {
+    return this._action;
+  }
+
+  set action(value: TimelineItemAction) {
+    this._action = value;
+  }
+
+  @IsObject()
+  private _createdAt: firebase.firestore.Timestamp;
+
+  get createdAt(): firestore.Timestamp {
     return this._createdAt;
   }
 
   set createdAt(value: firebase.firestore.Timestamp) {
     this._createdAt = value;
   }
 
-  // eslint-disable-next-line class-methods-use-this
-  fromFirestore(
-    data: firebase.firestore.QueryDocumentSnapshot<ITimelineItem>,
-  ): TimelineItem {
-    return TimelineItem.factory(data.data());
+  static factory(data: ITimelineItem): TimelineItem {
+    return new TimelineItem(
+      data.actorRef,
+      data.offerRef,
+      data.requestRef,
+      User.factory(data.actorSnapshot),
+      data.offerSnapshot ? Offer.factory(data.offerSnapshot) : null,
+      Request.factory(data.requestSnapshot),
+      data.action,
+      data.createdAt,
+    );
   }
 
-  // eslint-disable-next-line class-methods-use-this
-  toFirestore(modelObject: TimelineItem): ITimelineItem {
+  toObject(): object {
     return {
-      offerSnapshot: modelObject.offerSnapshot,
-      requestSnapshot: modelObject.requestSnapshot,
-      createdAt: modelObject._createdAt,
+      actorRef: this.actorRef.path,
+      offerRef: this.offerRef ? this.offerRef.path : null,
```

Comment:
@sabind can you review this?

I locally tested the web-client model by using that model to create a document in the database. It's not working without this change. It gets undefined if we use `this.offerRef?.path`. Without that, I was getting an `unsupported field offerRef` in the toFirebase method, because it returns undefined. Maybe we should refactor Request & User models too. 

I faced the same problem in other models as well. When createdAt is optional `createdAt?.toDate()` causes an error too.

Here's a simple example case: 
```javascript
class TimelineItem {
    public ref?: { path:string } | null = null;
    constructor(ref: { path:string} | null = null) {
        this.ref = ref;
    }
}

const item = new TimelineItem({ path: 'hello'});
console.log(item.ref ? item.ref.path : null); // prints 'hello'
console.log(item.ref?.path); // prints 'hello'

const itemWithoutRef = new TimelineItem();
console.log(itemWithoutRef.ref ? itemWithoutRef.ref.path : null); // prints 'hello'
console.log(itemWithoutRef.ref?.path); // prints undefined

````

Reply:
yes the conditional setting of the value as `null` is fine.